core/commands/dag/dag.go

```diff
@@ -96,56 +92,93 @@ into an object of the specified format.
 			}
 		}
 
-		if dopin {
-			defer n.Blockstore.PinLock().Unlock()
-		}
+		outChan := make(chan interface{}, 8)
+		res.SetOutput((<-chan interface{})(outChan))
 
-		nds, err := coredag.ParseInputs(ienc, format, fi, mhType, -1)
-		if err != nil {
-			res.SetError(err, cmds.ErrNormal)
-			return
-		}
-		if len(nds) == 0 {
-			res.SetError(fmt.Errorf("no node returned from ParseInputs"), cmds.ErrNormal)
-			return
-		}
+		addAllAndPin := func(f files.File) error {
+			for {
+				file, err := f.NextFile()
+				if err == io.EOF {
+					// Finished the list of files.
+					break
+				} else if err != nil {
+					return err
+				}
 
-		b := n.DAG.Batch()
-		for _, nd := range nds {
-			_, err := b.Add(nd)
-			if err != nil {
-				res.SetError(err, cmds.ErrNormal)
-				return
+				nds, err := coredag.ParseInputs(ienc, format, file, mhType, -1)
+				if err != nil {
+					return err
+				}
+				if len(nds) == 0 {
+					return fmt.Errorf("no node returned from ParseInputs")
+				}
+
+				var cid *cid.Cid
+				err = func() error {
+					if dopin {
+						defer n.Blockstore.PinLock().Unlock()
+					}
+
+					b := n.DAG.Batch()
```

Comment:
We should put all dag nodes being added into the same batch here, i would move this batch creation up above the start of the for loop.

Reply:
Are you sure it is right thing to do? What if I start adding 100GB of DAGs?

Reply:
That was my thinking there, though we might just say that users should care about their batch sizes.

Reply:
The DAG batch is size limited, it will autoflush if you put too many items into it

Reply:
done